File: Ordination/Python/QuickSort.py
#coding : utf- 8

from Sort import Sort

class QuickSort(Sort):
	pass

	def sort(self,array,leftIndex = 0 ,rightIndex = None):
		if(rightIndex == None):
			rightIndex = len(array) - 1
		#self.__simpleQuickSort(array,leftIndex,rightIndex)
		self.__quickSortMedianOfThree(array,leftIndex,rightIndex)


	def __quickSortMedianOfThree(self,array,leftIndex,rightIndex):
		if(self.__validateParams(array,leftIndex,rightIndex)):
			self.__medianOfThree(array,leftIndex,rightIndex)
			mid = (leftIndex + rightIndex ) / 2
			array[mid] , array[rightIndex - 1]  = array[rightIndex - 1],array[mid]
			partition  = self.__reversePartition(array,leftIndex,rightIndex)
			self.__quickSortMedianOfThree(array,leftIndex,partition -1)
			self.__quickSortMedianOfThree(array,partition + 1 , rightIndex)

	def __reversePartition(self,array,leftIndex,rightIndex):
		pivot = array[rightIndex]
		pivot_index = rightIndex
		for i in range(rightIndex, leftIndex - 1 , -1):
			if(pivot < array[i]):
				pivot_index = pivot_index - 1
				array[pivot_index] , array[i] = array[i] , array[pivot_index]
		array[pivot_index],array[rightIndex] = array[rightIndex],array[pivot_index]
		return pivot_index

	def __medianOfThree(self,array,leftIndex,rightIndex):
		mid = (leftIndex + rightIndex )  / 2
		minValue = leftIndex
		if(array[minValue] > array[rightIndex]):
			minValue = rightIndex
		if(array[mid] < array[minValue]):
			minValue = mid
		array[leftIndex],array[minValue] = array[minValue] , array[leftIndex]
		
		if(array[mid] > array[rightIndex]):
			array[mid] , array[rightIndex] = array[rightIndex] , array[mid]
		





	def __simpleQuickSort(self,array,leftIndex,rightIndex):
		if(self.__validateParams(array,leftIndex,rightIndex)):
			pivot = self.__simplePartition(array,leftIndex,rightIndex)
			self.__simpleQuickSort(array,leftIndex,pivot -1)
			self.__simpleQuickSort(array,pivot+1,rightIndex)
	
	def __simplePartition(self,array,leftIndex,rightIndex):
		pivot = array[leftIndex]
		pivot_index  = leftIndex
		for i in range(pivot_index + 1,rightIndex +1):
			if(pivot > array[i]):
				pivot_index = pivot_index + 1
				array[pivot_index] , array[i] = array[i] , array[pivot_index]
		array[leftIndex] ,array[pivot_index] = array[pivot_index],array[leftIndex]
		return pivot_index





	def __validateParams(self,array,leftIndex,rightIndex):
		if(array == None or leftIndex < 0 or rightIndex > len(array) - 1 or leftIndex >= rightIndex):
			return False
		else:
			return True


sort = QuickSort()
lista = [9,8,7,6,5,4,3,2,1]
sort.sort(lista,5,8)
print lista

File: Ordination/C++/QuickSort.cpp
#include <iostream>

template<typename T> class QuickSort{
	public:
		void sort(T array[],int leftIndex,int rightIndex,int size){
			if(validateParams(array,leftIndex,rightIndex,size)){
				//int partition = this->partition(array,leftIndex,rightIndex);
				int partition = this->recursivePartition(array,leftIndex,rightIndex);
				this->sort(array,leftIndex,partition - 1,size);
				this->sort(array,partition + 1,rightIndex,size);
			}else{
				//Do nothing
			}
		}
		
		void sort(T array[],int size){
			this->sort(array,0,size - 1 ,size);
			this->showArray(array,size);
		}
	private:
		int partition(T array[],int leftIndex,int rightIndex){
			T pivot = array[leftIndex];	
			int pivotIndex = leftIndex;
			for(int i = leftIndex +  1; i <= rightIndex;i++){
				if(pivot > array[i]){
					pivotIndex++;
					swap(array,pivotIndex,i);
				}
			}
			swap(array,leftIndex,pivotIndex);
			return pivotIndex;
		}
		
		int recursivePartition(T array[],int leftIndex,int rightIndex){
			int pivot_index = leftIndex;
			T pivot = array[leftIndex];
			pivot_index = this->recursiveLoopPartition(array,pivot,rightIndex,pivot_index + 1,pivot_index);
			swap(array,leftIndex,pivot_index);
			return pivot_index;
		}
		
		int recursiveLoopPartition(T array[],T pivot,int rightIndex,int index,int pivot_index){
			if(index <= rightIndex){
				if(pivot > array[index]){
					pivot_index++;
					swap(array,pivot_index,index);
				}
				return this->recursiveLoopPartition(array,pivot,rightIndex,index + 1,pivot_index);
			}else{
				return pivot_index;
			}
		}

		bool validateParams(T array[],int leftIndex,int rightIndex,int size){
			if(array == NULL || leftIndex < 0 || rightIndex > size - 1 || leftIndex >= rightIndex){
				return false;
			}else{
				return true;
			}
		}
		
		void swap(T array[],int i,int j){
			T aux = array[i];
			array[i] = array[j];
			array[j] = aux;
		}
		
		void showArray(T array[],int size){
			std::cout << "Array : [";
			for(int i = 0 ; i < size; i++){
				if(i != size - 1){
					std::cout << array[i] << ",";
				}else{
					std::cout << array[i] << "]";
				}
			}
			std::cout << std::endl;
		}

};
